Per-channel baseline

raghuramank100 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26623 Per channel fake quant
* **#26516 Per-channel baseline**

Differential Revision: [D17342622](https://our.internmc.facebook.com/intern/diff/D17342622/)

raghuramank10000 committed:
Per-channel baseline



Differential Revision: [D17342622](https://our.internmc.facebook.com/intern/diff/D17342622/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per-channel baseline"


Differential Revision: [D17342622](https://our.internmc.facebook.com/intern/diff/D17342622/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per-channel baseline"


Differential Revision: [D17342622](https://our.internmc.facebook.com/intern/diff/D17342622/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per-channel baseline"


Differential Revision: [D17342622](https://our.internmc.facebook.com/intern/diff/D17342622/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per-channel baseline"


Differential Revision: [D17342622](https://our.internmc.facebook.com/intern/diff/D17342622/)

[ghstack-poisoned]

jerryzh168 commented on test/test_quantized_models.py:
```diff
@@ -25,5 +25,26 @@ def test_float_quant_compare(self):
+    def test_float_quant_compare_per_channel(self):
+        # Test for per-channel Quant
+        torch.manual_seed(67)
+        myModel = ModelMultipleOps().to(torch.float32)
```
nit: can you use '_' naming e.g. 'my_model', or maybe just 'model'? so that we have a consistent naming?

jerryzh168 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -124,17 +128,20 @@ def forward(self, input):
-                                    self.stride, self.padding,
-                                    self.dilation, self.groups,
-                                    self.scale, self.zero_point)
+        else:
```
do we need this else?

dzhulgakov replied:
yeah, please revert?

jerryzh168 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -124,17 +128,20 @@ def forward(self, input):
     # regular QTensor form for serialization. Packed weights should not live
     # outside the process in which they were created, rather they should be derived
     # from the QTensor weight.
+
```
remove blank changes?

jerryzh168 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -106,7 +106,11 @@ def set_weight_bias(self, w, b):
         self._packed_params = torch.ops.quantized.conv_prepack(
             w, b, self.stride, self.padding, self.dilation, self.groups)
-        self.weight_scale = w.q_scale()
+        if w.qscheme() in set([torch.per_channel_affine, torch.per_channel_symmetric]):
```
maybe wrap this check into a function? I think we might need this check somewhere else as well

jerryzh168 commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -233,10 +240,15 @@ def from_float(cls, mod):
         act_scale, act_zp = activation_observer.calculate_qparams()
         assert weight_observer.dtype == torch.qint8, 'Weight observer must have a dtype of qint8'
         wt_scale, wt_zp = weight_observer.calculate_qparams()
+        if weight_observer.qscheme in set([torch.per_tensor_symmetric, torch.per_tensor_affine]):
```
same here

jerryzh168 commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -198,7 +198,11 @@ def bias(self):
         # type: (torch.Tensor, Optional[torch.Tensor]) -> None
         self._packed_params = torch.ops.quantized.linear_prepack(w, b)
-        self.weight_scale = w.q_scale()
+        if w.qscheme() in set([torch.per_channel_affine, torch.per_channel_symmetric]):
```
same here

jerryzh168 commented on torch/nn/quantized/modules/linear.py:
```diff
@@ -229,7 +233,14 @@ def from_float(cls, mod):
         assert weight_observer.dtype == torch.qint8, 'Weight observer must have dtype torch.qint8'
         wt_scale, wt_zp = weight_observer.calculate_qparams()
-        qweight = torch.quantize_linear(mod.weight.float(), float(wt_scale), int(wt_zp), torch.qint8)
+        if weight_observer.qscheme in set([torch.per_tensor_symmetric, torch.per_tensor_affine]):
```
and here

jerryzh168 replied:
also this block seems to be repeated as well, could you write another function that wraps around this?
e.g.
```
def quantize_weight(mod, weight_observer):
      wt_scale, wt_zp = weight_observer.calculate_qparams()
      if weight_observer.qscheme...
      return qweight
```

jerryzh168 reviewed (approved):
some changes requested, see inline comments

raghuramank10000 committed:
Update on "Per-channel baseline"


Differential Revision: [D17342622](https://our.internmc.facebook.com/intern/diff/D17342622/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per-channel baseline"


Differential Revision: [D17342622](https://our.internmc.facebook.com/intern/diff/D17342622/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per-channel baseline"


Differential Revision: [D17342622](https://our.internmc.facebook.com/intern/diff/D17342622/)

[ghstack-poisoned]

dzhulgakov commented on torch/nn/_intrinsic/qat/modules/conv_fused.py:
```diff
@@ -80,13 +80,17 @@ def reset_parameters(self):
     def enable_fake_quant(self):
-        self.observer.enable()
-        self.weight_fake_quant.enable()
+        self.observer.enable_fake_quant()
```
why do you need this? I guess the .apply should work recursively?

raghuramank100 replied:
Good catch, that's not needed, removed it.

dzhulgakov commented on torch/nn/quantized/modules/conv.py:
```diff
@@ -233,10 +235,14 @@ def from_float(cls, mod):
-        qweight = torch.quantize_per_tensor(
-            mod.weight.float(),
-            float(wt_scale), int(wt_zp), torch.qint8)
+        if weight_observer.qscheme in set([torch.per_tensor_symmetric, torch.per_tensor_affine]):
```
nit: `set([...])` -> `{...}`

raghuramank10000 committed:
Update on "Per-channel baseline"


Differential Revision: [D17342622](https://our.internmc.facebook.com/intern/diff/D17342622/)

[ghstack-poisoned]

jerryzh168 reviewed (changes requested):
looks like there are still some changes needs to be done. see dima's comments

raghuramank10000 committed:
Update on "Per-channel baseline"


Differential Revision: [D17342622](https://our.internmc.facebook.com/intern/diff/D17342622/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per-channel baseline"


Differential Revision: [D17342622](https://our.internmc.facebook.com/intern/diff/D17342622/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per-channel baseline"


Differential Revision: [D17342622](https://our.internmc.facebook.com/intern/diff/D17342622/)

[ghstack-poisoned]

dzhulgakov reviewed (approved):
Tests don't seem to pass here (observer scriptability) but do on the later PRs. You might need to do some reordering. Otherwise looks good.

raghuramank10000 committed:
Update on "Per-channel baseline"


Differential Revision: [D17342622](https://our.internmc.facebook.com/intern/diff/D17342622/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per-channel baseline"


Differential Revision: [D17342622](https://our.internmc.facebook.com/intern/diff/D17342622/)

[ghstack-poisoned]

jerryzh168 commented on torch/nn/quantized/modules/utils.py:
```diff
@@ -0,0 +1,13 @@
+
+def _quantize_weight(float_wt, observer):
+    wt_scale, wt_zp = observer.calculate_qparams()
+    if observer.qscheme in {torch.per_tensor_symmetric, torch.per_tensor_affine}:
```
nit: this can be a list

jerryzh168 commented on test/test_quantized_models.py:
```diff
@@ -4,26 +4,47 @@
 from common_quantization import QuantizationTestCase, ModelMultipleOps
```
why is this not in test_quantization.py?

jerryzh168 replied:
this looks like an integration test

raghuramank100 replied:
Let's chat. test_quantized_models covers tests at a higher level: more complex models and comparison of float with quantized numerics at the model level. I think its better to have it as a separate file as test_quantization is really large already

jerryzh168 reviewed (approved):
maybe remove `test_quantized_model.py` and move the code to `test_quantization.py`? could be in another PR

raghuramank10000 committed:
Update on "Per-channel baseline"


Differential Revision: [D17342622](https://our.internmc.facebook.com/intern/diff/D17342622/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per-channel baseline"


Differential Revision: [D17342622](https://our.internmc.facebook.com/intern/diff/D17342622/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per-channel baseline"


Differential Revision: [D17342622](https://our.internmc.facebook.com/intern/diff/D17342622/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per-channel baseline"


Differential Revision: [D17342622](https://our.internmc.facebook.com/intern/diff/D17342622/)

[ghstack-poisoned]

raghuramank10000 committed:
Update on "Per-channel baseline"


Differential Revision: [D17342622](https://our.internmc.facebook.com/intern/diff/D17342622/)

[ghstack-poisoned]
